File: packages/sdk/core/src/transport.ts
import type { ErrorEvent, PerformanceEvent, Transport, BugwatchOptions } from "./types";
import { createPersistentQueue, type PersistentQueue } from "./persistent-queue";

const DEFAULT_ENDPOINT = "https://api.bugwatch.dev";
const DEFAULT_TIMEOUT_MS = 10000;
const MAX_PAYLOAD_BYTES = 512 * 1024; // 512 KB
const MAX_RETRIES = 2;
const BASE_RETRY_DELAY_MS = 500;
const DRAIN_RATE_DELAY_MS = 200; // 5 events/sec on drain to avoid thundering herd

/**
 * Safely serialize a value to JSON, handling circular references.
 */
function safeJsonStringify(value: unknown): string {
  const seen = new WeakSet();
  return JSON.stringify(value, (_key, val) => {
    if (typeof val === "object" && val !== null) {
      if (seen.has(val)) {
        return "[Circular]";
      }
      seen.add(val);
    }
    return val;
  });
}

/**
 * Truncate an event's breadcrumbs and extra data to fit within size limits.
 * Returns the serialized JSON string.
 */
function serializeWithTruncation(event: ErrorEvent): string {
  let json = safeJsonStringify(event);

  if (json.length <= MAX_PAYLOAD_BYTES) {
    return json;
  }

  // First pass: drop breadcrumbs progressively
  if (event.breadcrumbs && event.breadcrumbs.length > 0) {
    const truncated = { ...event, breadcrumbs: [] as typeof event.breadcrumbs };
    json = safeJsonStringify(truncated);

    if (json.length <= MAX_PAYLOAD_BYTES) {
      return json;
    }
  }

  // Second pass: drop extra data and breadcrumbs
  const stripped = { ...event, breadcrumbs: undefined, extra: undefined };
  json = safeJsonStringify(stripped);

  if (json.length <= MAX_PAYLOAD_BYTES) {
    return json;
  }

  // Third pass: truncate exception stacktrace
  if (stripped.exception?.stacktrace && stripped.exception.stacktrace.length > 5) {
    stripped.exception = {
      ...stripped.exception,
      stacktrace: stripped.exception.stacktrace.slice(0, 5),
    };
    json = safeJsonStringify(stripped);
  }

  // Final pass: truncate message if still too large
  if (json.length > MAX_PAYLOAD_BYTES && stripped.message && stripped.message.length > 500) {
    stripped.message = stripped.message.slice(0, 500) + "...(truncated)";
    json = safeJsonStringify(stripped);
  }

  return json;
}

/**
 * Check if an HTTP status code is retryable
 */
function isRetryableStatus(status: number): boolean {
  return status === 429 || status >= 500;
}

/**
 * HTTP transport for sending events to the Bugwatch API
 */
export class HttpTransport implements Transport {
  private endpoint: string;
  private apiKey: string;
  private debug: boolean;
  private timeoutMs: number;
  private inFlightRequests: Set<Promise<void>> = new Set();
  private rateLimitedUntil: number = 0;
  private persistentQueue: PersistentQueue;
  private drainPromise: Promise<void> | null = null;
  private onDropped?: BugwatchOptions["onDropped"];

  constructor(options: BugwatchOptions) {
    this.endpoint = options.endpoint || DEFAULT_ENDPOINT;
    this.apiKey = options.apiKey;
    this.debug = options.debug || false;
    this.timeoutMs = DEFAULT_TIMEOUT_MS;
    this.onDropped = options.onDropped;
    this.persistentQueue = createPersistentQueue(options.offline);
    // Drain on init — fire and forget. Failures are silent.
    this.drainPersistentQueue();
  }

  /**
   * Drain queued events from previous sessions / network outages.
   * Rate-limited to avoid thundering herd on a freshly recovered connection.
   */
  private drainPersistentQueue(): void {
    if (this.drainPromise) return;
    this.drainPromise = (async () => {
      try {
        const events = await this.persistentQueue.drain();
        if (events.length === 0) return;
        if (this.debug) {
          console.log(`[Bugwatch] Draining ${events.length} queued event(s)`);
        }
        for (const event of events) {
          // Re-send via the standard path, but skip persistent queue on failure
          // to avoid an infinite re-queue loop.
          await this.sendWithRetry(event, 0, /* persistOnFailure */ false).catch(() => undefined);
          await new Promise((r) => setTimeout(r, DRAIN_RATE_DELAY_MS));
        }
      } finally {
        this.drainPromise = null;
      }
    })();
  }

  async send(event: ErrorEvent): Promise<void> {
    // Respect rate limit backoff
    if (Date.now() < this.rateLimitedUntil) {
      if (this.debug) {
        console.warn("[Bugwatch] Rate limited, dropping event:", event.event_id);
      }
      return;
    }

    // Wrap in a new promise so we can add it to the Set synchronously
    // before any async work starts, preventing a race with flush().
    let resolve!: () => void;
    const tracked = new Promise<void>((r) => { resolve = r; });
    this.inFlightRequests.add(tracked);
    this.sendWithRetry(event).finally(() => {
      this.inFlightRequests.delete(tracked);
      resolve();
    });
  }

  private async sendWithRetry(event: ErrorEvent, attempt = 0, persistOnFailure = true): Promise<void> {
    const url = `${this.endpoint}/api/v1/events`;

    if (this.debug && attempt === 0) {
      console.log("[Bugwatch] Sending event:", event.event_id);
    }

    try {
      const controller = typeof AbortController !== "undefined" ? new AbortController() : null;
      const timeoutId = controller
        ? setTimeout(() => controller.abort(), this.timeoutMs)
        : null;

      try {
        const body = serializeWithTruncation(event);

        const response = await this.fetch(url, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            "Authorization": `Bearer ${this.apiKey}`,
            "X-Bugwatch-SDK": event.sdk?.name || "bugwatch-core",
            "X-Bugwatch-SDK-Version": event.sdk?.version || "0.1.0",
          },
          body,
          ...(controller && { signal: controller.signal }),
        });

        if (!response.ok) {
          // Handle rate limiting with backoff
          if (response.status === 429) {
            const retryAfter = response.headers.get("Retry-After");
            const parsed = retryAfter ? parseInt(retryAfter, 10) : NaN;
            const backoffMs = !isNaN(parsed) && parsed > 0
              ? parsed * 1000
              : 60_000;
            this.rateLimitedUntil = Date.now() + backoffMs;

            if (this.debug) {
              console.warn(`[Bugwatch] Rate limited. Backing off for ${backoffMs}ms`);
            }
            return; // Drop event, don't retry during rate limit
          }

          if (isRetryableStatus(response.status) && attempt < MAX_RETRIES) {
            const delay = BASE_RETRY_DELAY_MS * Math.pow(2, attempt) + Math.random() * 200;
            await new Promise((resolve) => setTimeout(resolve, delay));
            return this.sendWithRetry(event, attempt + 1);
          }

          if (this.debug) {
            const errorText = await response.text().catch(() => "");
            console.error("[Bugwatch] Failed to send event:", response.status, errorText);
          }
          return;
        }

        if (this.debug) {
          console.log("[Bugwatch] Event sent successfully:", event.event_id);
        }
      } finally {
        if (timeoutId !== null) {
          clearTimeout(timeoutId);
        }
      }
    } catch (error) {
      // Retry on network errors (timeout, connection refused, etc.)
      if (attempt < MAX_RETRIES) {
        const delay = BASE_RETRY_DELAY_MS * Math.pow(2, attempt) + Math.random() * 200;
        await new Promise((resolve) => setTimeout(resolve, delay));
        return this.sendWithRetry(event, attempt + 1, persistOnFailure);
      }

      if (this.debug) {
        const message = error instanceof Error && error.name === "AbortError"
          ? `[Bugwatch] Request timed out after ${this.timeoutMs}ms`
          : "[Bugwatch] Transport error:";
        console.error(message, error);
      }

      // Network failure after all retries — persist for next session if enabled
      if (persistOnFailure) {
        try {
          const ok = await this.persistentQueue.enqueue(event);
          if (ok && this.debug) {
            console.log("[Bugwatch] Event persisted to offline queue:", event.event_id);
          }
        } catch {
          // Queue itself failed — silent drop, never crash the host app
        }
      }
      // Don't throw - we don't want SDK errors to break the application
    }
  }

  /**
   * Send a performance transaction to the API.
   */
  async sendTransaction(event: PerformanceEvent): Promise<void> {
    if (Date.now() < this.rateLimitedUntil) {
      if (this.debug) {
        console.warn("[Bugwatch] Rate limited, dropping transaction:", event.transaction_name);
      }
      return;
    }

    const promise = (async () => {
      const url = `${this.endpoint}/api/v1/transactions`;

      if (this.debug) {
        console.log("[Bugwatch] Sending transaction:", event.transaction_name);
      }

      try {
        const controller = typeof AbortController !== "undefined" ? new AbortController() : null;
        const timeoutId = controller
          ? setTimeout(() => controller.abort(), this.timeoutMs)
          : null;

        try {
          const response = await this.fetch(url, {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
              "Authorization": `Bearer ${this.apiKey}`,
              "X-Bugwatch-SDK": "bugwatch-core",
            },
            body: JSON.stringify(event),
            ...(controller && { signal: controller.signal }),
          });

          if (!response.ok) {
            if (response.status === 429) {
              const retryAfter = response.headers.get("Retry-After");
              const parsed = retryAfter ? parseInt(retryAfter, 10) : NaN;
              const backoffMs = !isNaN(parsed) && parsed > 0 ? parsed * 1000 : 60_000;
              this.rateLimitedUntil = Date.now() + backoffMs;
            }
            if (this.debug) {
              console.error("[Bugwatch] Failed to send transaction:", response.status);
            }
            return;
          }

          if (this.debug) {
            console.log("[Bugwatch] Transaction sent successfully:", event.transaction_name);
          }
        } finally {
          if (timeoutId !== null) {
            clearTimeout(timeoutId);
          }
        }
      } catch (error) {
        if (this.debug) {
          console.error("[Bugwatch] Transport error sending transaction:", error);
        }
      }
    })();
    this.inFlightRequests.add(promise);
    promise.finally(() => this.inFlightRequests.delete(promise));
    await promise;
  }

  /**
   * Wait for all in-flight requests to complete.
   */
  async flush(): Promise<void> {
    if (this.inFlightRequests.size > 0) {
      await Promise.allSettled([...this.inFlightRequests]);
    }
  }

  /**
   * Abstract fetch to support different environments
   */
  private async fetch(
    url: string,
    options: RequestInit
  ): Promise<Response> {
    // Use global fetch (available in Node 18+, browsers, and edge runtimes)
    if (typeof globalThis.fetch === "function") {
      return globalThis.fetch(url, options);
    }

    throw new Error("No fetch implementation available");
  }
}

/**
 * No-op transport for testing or disabled SDK
 */
export class NoopTransport implements Transport {
  async send(_event: ErrorEvent): Promise<void> {
    // Do nothing
  }
}

/**
 * Console transport for development/debugging
 */
export class ConsoleTransport implements Transport {
  async send(event: ErrorEvent): Promise<void> {
    console.log("[Bugwatch Event]", safeJsonStringify(event));
  }
}

/**
 * Batching transport that queues events and sends them in batches
 */
export class BatchTransport implements Transport {
  private transport: Transport;
  private queue: ErrorEvent[] = [];
  private maxBatchSize: number;
  private flushInterval: number;
  private timer: ReturnType<typeof setInterval> | null = null;

  constructor(
    transport: Transport,
    options: { maxBatchSize?: number; flushInterval?: number } = {}
  ) {
    this.transport = transport;
    this.maxBatchSize = options.maxBatchSize || 10;
    this.flushInterval = options.flushInterval || 5000;

    // Start flush timer
    this.startTimer();
  }

  async send(event: ErrorEvent): Promise<void> {
    this.queue.push(event);

    if (this.queue.length >= this.maxBatchSize) {
      await this.flush();
    }
  }

  async flush(): Promise<void> {
    if (this.queue.length === 0) {
      // Also flush the underlying transport's in-flight requests
      if (this.transport.flush) {
        await this.transport.flush();
      }
      return;
    }

    const events = this.queue.splice(0, this.maxBatchSize);

    // Send events in parallel, don't fail the batch if individual events fail
    await Promise.allSettled(events.map((event) => this.transport.send(event)));

    // If there are still events in the queue, flush recursively
    if (this.queue.length > 0) {
      await this.flush();
    }

    // Wait for underlying transport's in-flight requests
    if (this.transport.flush) {
      await this.transport.flush();
    }
  }

  private startTimer(): void {
    if (typeof setInterval !== "undefined") {
      this.timer = setInterval(() => {
        this.flush().catch(() => {
          // Ignore flush errors
        });
      }, this.flushInterval);

      // Unref timer in Node.js to not block process exit
      if (this.timer && typeof this.timer === "object" && "unref" in this.timer) {
        (this.timer as NodeJS.Timeout).unref();
      }
    }
  }

  destroy(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  /**
   * Close the transport, flushing any remaining events and stopping the timer.
   */
  async close(): Promise<void> {
    this.destroy(); // Stop timer first to prevent new flushes
    await this.flush();
  }
}

File: packages/sdk/rust/src/transport.rs
//! HTTP transport for sending error events to Bugwatch.

use std::time::Duration;
use thiserror::Error;

use crate::types::{BugwatchOptions, ErrorEvent};

const MAX_RETRIES: u32 = 2;
const BASE_RETRY_DELAY_MS: u64 = 500;

fn is_retryable_status(status: u16) -> bool {
    status == 429 || status >= 500
}

/// Transport errors.
#[derive(Error, Debug)]
pub enum TransportError {
    #[error("HTTP error: {0}")]
    Http(String),
    #[error("Serialization error: {0}")]
    Serialization(#[from] serde_json::Error),
    #[error("Request error: {0}")]
    Request(String),
}

/// Trait for sending events to Bugwatch.
pub trait Transport: Send + Sync {
    /// Send an event to Bugwatch.
    fn send(&self, event: &ErrorEvent) -> Result<(), TransportError>;

    /// Flush any pending events.
    ///
    /// This ensures all queued events are sent before the method returns.
    /// The default implementation does nothing (for transports that send synchronously).
    fn flush(&self) -> Result<(), TransportError> {
        Ok(())
    }

    /// Close the transport and release any resources.
    ///
    /// After calling this method, the transport should not be used again.
    /// The default implementation calls flush().
    fn close(&self) -> Result<(), TransportError> {
        self.flush()
    }

    /// Downcast support for accessing concrete transport types.
    ///
    /// The default implementation returns a reference that will fail any downcast,
    /// causing the client to fall back to synchronous `send()`. Override this in
    /// custom transports that need async support.
    fn as_any(&self) -> &dyn std::any::Any {
        &()
    }
}

/// HTTP transport for sending events.
#[derive(Clone)]
pub struct HttpTransport {
    endpoint: String,
    api_key: String,
    debug: bool,
    #[cfg(feature = "blocking")]
    blocking_client: Option<reqwest::blocking::Client>,
    #[cfg(feature = "async")]
    async_client: reqwest::Client,
}

impl HttpTransport {
    /// Create a new HTTP transport.
    ///
    /// # Errors
    ///
    /// Returns `TransportError::Request` if the HTTP client cannot be created.
    pub fn new(options: &BugwatchOptions) -> Result<Self, TransportError> {
        #[cfg(feature = "blocking")]
        let blocking_client = Some(
            reqwest::blocking::Client::builder()
                .timeout(Duration::from_secs(10))
                .build()
                .map_err(|e| {
                    TransportError::Request(format!("Failed to create HTTP client: {}", e))
                })?,
        );

        #[cfg(feature = "async")]
        let async_client = reqwest::Client::builder()
            .timeout(Duration::from_secs(10))
            .build()
            .map_err(|e| {
                TransportError::Request(format!("Failed to create async HTTP client: {}", e))
            })?;

        Ok(Self {
            endpoint: format!("{}/api/v1/events", options.endpoint.trim_end_matches('/')),
            api_key: options.api_key.clone(),
            debug: options.debug,
            #[cfg(feature = "blocking")]
            blocking_client,
            #[cfg(feature = "async")]
            async_client,
        })
    }

    /// Send an event synchronously (blocking) with retry.
    #[cfg(feature = "blocking")]
    pub fn send_blocking(&self, event: &ErrorEvent) -> Result<(), TransportError> {
        let client = self
            .blocking_client
            .as_ref()
            .ok_or_else(|| TransportError::Request("No blocking client available".to_string()))?;

        // Serialize explicitly so we get a clear serialization error vs network error
        let body = serde_json::to_vec(event)?;

        let mut last_err = None;
        for attempt in 0..=MAX_RETRIES {
            if attempt > 0 {
                std::thread::sleep(Duration::from_millis(
                    BASE_RETRY_DELAY_MS * (1 << (attempt - 1)),
                ));
            }
            match client
                .post(&self.endpoint)
                .header("Content-Type", "application/json")
                .header("Authorization", format!("Bearer {}", self.api_key))
                .header(
                    "User-Agent",
                    format!("bugwatch-rust/{}", env!("CARGO_PKG_VERSION")),
                )
                .body(body.clone())
                .send()
            {
                Ok(response) => {
                    let status = response.status().as_u16();
                    if response.status().is_success() {
                        if self.debug {
                            tracing::debug!("Event sent successfully: {}", event.event_id);
                        }
                        return Ok(());
                    } else if is_retryable_status(status) {
                        last_err = Some(TransportError::Http(format!("HTTP {}", status)));
                        continue;
                    } else {
                        return Err(TransportError::Http(format!("HTTP {}", status)));
                    }
                }
                Err(e) => {
                    last_err = Some(TransportError::Request(e.to_string()));
                    continue;
                }
            }
        }
        Err(last_err.unwrap_or_else(|| TransportError::Request("Unknown error".to_string())))
    }

    /// Send an event asynchronously with retry.
    #[cfg(feature = "async")]
    pub async fn send_async(&self, event: &ErrorEvent) -> Result<(), TransportError> {
        // Serialize explicitly so we get a clear serialization error vs network error
        let body = serde_json::to_vec(event)?;

        let mut last_err = None;
        for attempt in 0..=MAX_RETRIES {
            if attempt > 0 {
                tokio::time::sleep(Duration::from_millis(
                    BASE_RETRY_DELAY_MS * (1 << (attempt - 1)),
                ))
                .await;
            }
            match self
                .async_client
                .post(&self.endpoint)
                .header("Content-Type", "application/json")
                .header("Authorization", format!("Bearer {}", self.api_key))
                .header(
                    "User-Agent",
                    format!("bugwatch-rust/{}", env!("CARGO_PKG_VERSION")),
                )
                .body(body.clone())
                .send()
                .await
            {
                Ok(response) => {
                    let status = response.status().as_u16();
                    if response.status().is_success() {
                        if self.debug {
                            tracing::debug!("Event sent successfully: {}", event.event_id);
                        }
                        return Ok(());
                    } else if is_retryable_status(status) {
                        last_err = Some(TransportError::Http(format!("HTTP {}", status)));
                        continue;
                    } else {
                        return Err(TransportError::Http(format!("HTTP {}", status)));
                    }
                }
                Err(e) => {
                    last_err = Some(TransportError::Request(e.to_string()));
                    continue;
                }
            }
        }
        Err(last_err.unwrap_or_else(|| TransportError::Request("Unknown error".to_string())))
    }
}

impl Transport for HttpTransport {
    fn send(&self, event: &ErrorEvent) -> Result<(), TransportError> {
        #[cfg(feature = "blocking")]
        {
            self.send_blocking(event)
        }
        #[cfg(not(feature = "blocking"))]
        {
            Err(TransportError::Request(
                "Blocking transport not available. Enable the 'blocking' feature.".to_string(),
            ))
        }
    }

    fn as_any(&self) -> &dyn std::any::Any {
        self
    }
}

/// No-op transport that does nothing (for testing).
#[derive(Default)]
pub struct NoopTransport;

impl Transport for NoopTransport {
    fn send(&self, _event: &ErrorEvent) -> Result<(), TransportError> {
        Ok(())
    }

    fn as_any(&self) -> &dyn std::any::Any {
        self
    }
}

/// Console transport that logs events (for debugging).
pub struct ConsoleTransport {
    debug: bool,
}

impl ConsoleTransport {
    pub fn new(debug: bool) -> Self {
        Self { debug }
    }
}

impl Transport for ConsoleTransport {
    fn send(&self, event: &ErrorEvent) -> Result<(), TransportError> {
        println!("[Bugwatch] Event {}", event.event_id);
        println!("  Level: {}", event.level);
        if let Some(ref exception) = event.exception {
            println!("  Exception: {}: {}", exception.error_type, exception.value);
        }
        if let Some(ref message) = event.message {
            println!("  Message: {}", message);
        }
        println!("  Tags: {:?}", event.tags);
        Ok(())
    }

    fn as_any(&self) -> &dyn std::any::Any {
        self
    }
}

impl Default for ConsoleTransport {
    fn default() -> Self {
        Self::new(false)
    }
}
